feat: :sparkles: add configurable DNS addresses #30

pizokh opened the pull request:
- Added default DNS addresses and a command-line option (--dns-addr) to customize them.

markpash commented on app/app.go:
```diff
@@ -22,6 +22,7 @@ type WarpOptions struct {
 	Bind     netip.AddrPort
 	Endpoint string
 	License  string
+	DnsAddr  []string
```
Change this from `[]string` to `netip.Addr`.

markpash commented on main.go:
```diff
@@ -73,6 +88,7 @@ func main() {
 		bind     = fs.String('b', "bind", "127.0.0.1:8086", "socks bind address")
 		endpoint = fs.String('e', "endpoint", "", "warp endpoint")
 		key      = fs.String('k', "key", "", "warp key")
+		dnsAddr  = fs.StringList('d', "dns-addr", fmt.Sprintf("DNS addresses (default: %v)", defaultDnsAddr))
```
Let's change this to be a single string, using `fs.StringLong` and that should allow you to specify the default directly here. Let's just make the default `1.1.1.1`.

markpash commented on main.go:
```diff
@@ -62,6 +62,21 @@ var psiphonCountries = []string{
 	"US",
 }
 
+var defaultDnsAddr = []string{
```
Let's just get rid of this list, it causes more confusion for users. We'll just use `1.1.1.1` as the default.

markpash commented on main.go:
```diff
@@ -131,10 +147,15 @@ func main() {
 		fatal(l, fmt.Errorf("invalid bind address: %w", err))
 	}
 
+	if len(*dnsAddr) == 0 {
```
This won't be needed if the above comments are implemented.

markpash commented on warp/account.go:
```diff
@@ -261,24 +261,21 @@ func updateLicenseKey(accountID, accessToken, license string) (IdentityAccount,
-	buffer.WriteString("2620:fe::fe, ")
-	buffer.WriteString("2620:fe::9\n")
+
+	for index, server := range dnsAddr {
```
Instead of writing out the DNS here, what we should do is remove it from here, and do it in each of the `runWarp(), runWarpInWarp() and runWarpWithPsiphon()` functions in `app.go`. Just like how `conf.Interface.MTU` is set, we should have a `conf.Interface.DNS`.

markpash reviewed (commented):
Good Job! I've left some comments for some changes.

pizokh commented:
Thank you for the feedback, @markpash! I have made the requested changes:

- Changed `DnsAddr` from `[]string` to `netip.Addr` in `WarpOptions`.
- Updated the command-line option to be a single string using `fs.StringLong` with `1.1.1.1` as the default.
- Removed the `defaultDnsAddr` list and associated logic.
- Adjusted the DNS setup to be handled directly in `runWarp()`, `runWarpInWarp()`, and `runWarpWithPsiphon()` functions.
- Updated the `createConf` and `LoadOrCreateIdentity` functions to reflect these changes.

Please review the updated implementation. Let me know if any further adjustments are needed.

markpash commented on wiresocks/config.go:
```diff
@@ -78,20 +78,17 @@ func ParseInterface(cfg *ini.File) (InterfaceConfig, error) {
+			if err != nil {
+				return InterfaceConfig{}, err
+			}
+			addresses = append(addresses, ip)
```
You probably shouldn't use `append()` when you know the size of the resulting slice beforehand.
Consider the following:
```go
addrs := sectionKey.StringsWithShadows(",")
device.DNS = make([]netip.Addr, len(addrs))
for i, addr := range addrs {
    ip, err := netip.ParseAddr(str)
    if err != nil {
        return InterfaceConfig{}, err
    }
    device.DNS[i] = ip
}
```

markpash commented on main.go:
```diff
@@ -73,6 +73,7 @@ func main() {
 		bind     = fs.String('b', "bind", "127.0.0.1:8086", "socks bind address")
 		endpoint = fs.String('e', "endpoint", "", "warp endpoint")
 		key      = fs.String('k', "key", "", "warp key")
+		dns      = fs.StringLong("dns-addr", "1.1.1.1", "DNS address")
```
Let's just make this `dns` instead of `dns-addr`.
Also please don't forget to add the line to the example config json file.

markpash replied:
Also don't forget to edit your commit message as well.
